src/v/datalake/tests/proto_to_arrow_test_utils.h

```diff
@@ -0,0 +1,370 @@
+            random_generators::get_int<int16_t>(0, allow_compression ? 4 : 0)),
+          .last_offset_delta = num_records - 1,
+          .first_timestamp = ts,
+          .max_timestamp = max_ts,
+          .producer_id = -1,
+          .producer_epoch = -1,
+          .base_sequence = 0,
+          .record_count = num_records};
+
+        auto size = model::packed_record_batch_header_size;
+        model::record_batch::records_type records;
+        auto rs = model::record_batch::uncompressed_records();
+        rs.reserve(num_records);
+        for (int i = 0; i < num_records; i++) {
+            std::stringstream key_stream;
+            key_stream << i;
+
+            std::string value = generate_simple_message(
+              "message # " + key_stream.str(), i);
+            rs.emplace_back(
+              make_record(offset() + i, ts() + i, key_stream.str(), value));
+        }
+
+        if (header.attrs.compression() != model::compression::none) {
+            iobuf body;
+            for (auto& r : rs) {
+                model::append_record_to_buffer(body, r);
+            }
+            rs.clear();
+            records = compression::compressor::compress(
+              body, header.attrs.compression());
+            size += std::get<iobuf>(records).size_bytes();
+        } else {
+            for (auto& r : rs) {
+                size += r.size_bytes();
+                size += vint::vint_size(r.size_bytes());
+            }
+            records = std::move(rs);
+        }
+        header.ctx = model::record_batch_header::context(
+          model::term_id(0), ss::this_shard_id());
+        header.size_bytes = size;
+        auto batch = model::record_batch(header, std::move(records));
+        batch.header().crc = model::crc_record_batch(batch);
+        batch.header().header_crc = model::internal_header_only_crc(
+          batch.header());
+        return batch;
+    }
+
+    model::record make_record(
+      int ts_delta,
+      int offset_delta,
+      std::string key_string,
+      std::string value_string) {
+        iobuf key;
+        iobuf value;
+        key.append(key_string.data(), key_string.size());
+        value.append(value_string.data(), value_string.size());
+        // Headers in the storages tests are just random. We can leave them
+        // empty.
+        std::vector<model::record_header> headers;
+
+        auto size = sizeof(model::record_attributes::type) // attributes
+                    + vint::vint_size(ts_delta)            // timestamp delta
+                    + vint::vint_size(offset_delta)        // offset delta
+                    + vint::vint_size(key.size_bytes())    // size of key-len
+                    + key.size_bytes()                     // size of key
+                    + vint::vint_size(value.size_bytes())  // size of value
+                    + value.size_bytes() // size of value (includes lengths)
+                    + vint::vint_size(headers.size());
```

Comment:
We should not be copying this out. I think there is a record builder somewhere for this kind of thing, let me look...

Reply:
Looks like people tend to use storage::batch_builder: https://github.com/redpanda-data/redpanda/blob/78e502348eecaa896425dd4d6816742a706209e5/src/v/wasm/tests/wasm_fixture.cc#L230-L233

Can we do that here? I'll try to find time to move that out of storage...

Reply:
👍